File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/K.Сортировка_слиянием/solution.py
def merge(arr: list[int], lf: int, mid: int, rg: int) -> list[int]:
    left_part: list[int] = arr[lf:mid]
    right_part: list[int] = arr[mid:rg]
    main_idx: int = lf
    left: int = 0
    right: int = 0
    while left < len(left_part) and right < len(right_part):
        if left_part[left] <= right_part[right]:
            arr[main_idx] = left_part[left]
            left += 1
        else:
            arr[main_idx] = right_part[right]
            right += 1
        main_idx += 1
    while left < len(left_part):
        arr[main_idx] = left_part[left]
        left += 1
        main_idx += 1
    while right < len(right_part):
        arr[main_idx] = right_part[right]
        right += 1
        main_idx += 1
    return arr



def merge_sort(arr, lf, rg):
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    if lf < rg - 1:
        mid: int = (lf + rg) // 2
        merge_sort(arr, lf, mid)
        merge_sort(arr, mid, rg)
        merge(arr, lf, mid, rg)

def test():
	a = [1, 4, 9, 2, 10, 11]
	b = merge(a, 0, 3, 6)
	expected = [1, 2, 4, 9, 10, 11]
	assert b == expected
	c = [1, 4, 2, 10, 1, 2]
	merge_sort(c, 0 , 6)
	expected = [1, 1, 2, 2, 4, 10]
	assert c == expected

if __name__ == '__main__':
    test()

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/K.Сортировка_слиянием/solution.cpp
#include <vector>
#include <cassert>

using Iterator = std::vector<int>::iterator;
using CIterator = std::vector<int>::const_iterator;

std::vector<int> merge(
    CIterator left_begin, CIterator left_end,
    CIterator right_begin, CIterator right_end) {

    std::vector<int> merged;
    CIterator left_it = left_begin;
    CIterator right_it = right_begin;
    while (left_it != left_end && right_it != right_end) {
        if (*left_it <= *right_it) {
            merged.push_back(*left_it);
            ++left_it;
        } else {
            merged.push_back(*right_it);
            ++right_it;
        }
    }
    while (left_it != left_end) {
        merged.push_back(*left_it);
        ++left_it;
    }
    while (right_it != right_end) {
        merged.push_back(*right_it);
        ++right_it;
    }

    return merged;
}

void merge_sort(Iterator begin, Iterator end) {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    if (end - begin > 1) {
        Iterator mid = begin + (end - begin) / 2;
        merge_sort(begin, mid);
        merge_sort(mid, end);
        std::vector<int> merged = merge(begin, mid, mid, end);
        std::copy(merged.begin(), merged.end(), begin);
    }
}

void test_merge_sort() {
	std::vector<int> a = {1, 4, 9};
	std::vector<int> b = {2, 10, 11};
	std::vector<int> c = merge(a.cbegin(), a.cend(), b.cbegin(), b.cend());
	std::vector<int> expected = {1, 2, 4, 9, 10, 11};
	assert(c == expected);
	std::vector<int> d = {1, 4, 2, 10, 1, 2};
	std::vector<int> sorted = {1, 1, 2, 2, 4, 10};
	merge_sort(d.begin(), d.end());
	assert(d == sorted);
}
